### PR TITLE
Add logic for VFs mac restoration to entrypoint.sh

### PR DESCRIPTION
**Pull request by ykulazhenkov**
Add additional logic to entrypoint.sh for Ubuntu, which sets
administrative mac address for VFs. After mac set, VF
will be unbinded from the driver and then binded again.
These actions are required to force GUID generation for
RDMA device.

### CODE REVIEW

File: ubuntu/entrypoint.sh [hunk 1]
```diff
@@ -80,6 +84,84 @@ function unload_modules() {
...
+    ip link set dev "$1" up
+}
+
+function driver_bind_vf() {
+    echo "$1" >$DRIVER_PATH/bind
+}
+
+function driver_unbind_vf() {
+    echo "$1" >$DRIVER_PATH/unbind
+}
```
Comment (moshe010) [inline ubuntu/entrypoint.sh:121]: you can create a method for driver_rebind_vf() 
driver_rebind_vf() 
{
    driver_bind_vf()
   driver_unbind_vf()
}

File: ubuntu/entrypoint.sh [hunk 2]
```diff
@@ -80,6 +84,84 @@ function unload_modules() {
...
+                vf_pci_addr=$(basename $(readlink $vf/device)) && [[ -n $vf_pci_addr ]] || return 1
+
+                echo "$pf_name" "$vf_name" "$vf_index" "$vf_mac" "$vf_pci_addr"
+            done
+        fi
+    done
+}
+
+function set_administrative_mac_for_vf() {
+    ip link set dev "$1" vf "$2" mac "$3"
```
Comment (moshe010) [inline ubuntu/entrypoint.sh:108]: general comment it better to assign the $1 $2 .. to local variable for readability

File: ubuntu/entrypoint.sh [hunk 3]
```diff
@@ -80,6 +84,84 @@ function unload_modules() {
...
+        pf_arr[$pf_name]=1
+        local vf_name=${vf_info[1]}
+        local vf_index=${vf_info[2]}
+        local vf_mac=${vf_info[3]}
+        local vf_pci_addr=${vf_info[4]}
+
+        echo "fix guid for VF: $vf_name"
+
+        if ! set_administrative_mac_for_vf "$pf_name" "$vf_index" "$vf_mac"; then
+            echo "failed to set mac for  VF $vf_index"
```
Comment (moshe010) [inline ubuntu/entrypoint.sh:144]: you have double space also you should specify the $pf_name

File: ubuntu/entrypoint.sh [hunk 4]
```diff
@@ -80,6 +84,84 @@ function unload_modules() {
...
+            echo "failed to unbind $pf_name VF $vf_index, $vf_pci_addr"
+            return 1
+        fi
+        if ! driver_bind_vf "$vf_pci_addr"; then
+            echo "failed to bind $pf_name VF $vf_index, $vf_pci_addr"
+            return 1
+        fi
+    done <<<"$vf_data"
+    local pf
+    for pf in "${!pf_arr[@]}";do
```
Comment (moshe010) [inline ubuntu/entrypoint.sh:157]: is this to solve the other bug opened in QA? if so I prefer it in another PR comment.
If not explain why it need to fix the rest mac issue

**Review by moshe010 — CHANGES_REQUESTED** [review]

**Comment by moshe010** [comment]
@yehorov-nvidia PTAL on this PR

File: ubuntu/entrypoint.sh [hunk 5]
```diff
@@ -91,6 +173,7 @@ fi
 unload_modules rpcrdma rdma_cm
 exit_on_error start_driver
 mount_rootfs
+fix_guid_for_vfs
```
Comment (adrianchiris) [inline ubuntu/entrypoint.sh:176]: nit: please add a comment explaining why we need to fix this for posterity

File: ubuntu/entrypoint.sh (hunk 4 above)
Comment (adrianchiris) [reply ubuntu/entrypoint.sh:157]: this bit sets link up for all mlnx PFs ? MOFED already does that  IIRC ? 
if its a bug and we need to explicitly set this up here , you should explain in a comment.

Also i would just set link up only to what was up before the driver has reloaded.

+1 on handling this bit in a separate PR

File: ubuntu/entrypoint.sh [hunk 6]
```diff
@@ -80,6 +84,84 @@ function unload_modules() {
...
 }
 
+
+# find all available vfs
+# return format <pf_name> <vf_name> <vf_index> <vf_mac> <vf_pci_addr>
+function find_mlx_vfs() {
+    for ethpath in /sys/class/net/*; do
+        if (grep $VENDOR "$ethpath"/device/vendor >/dev/null 2>&1); then
+            for vf in "$ethpath"/device/virtfn[0-9]*/net/*; do
+                [[ -d $vf ]] && [[ -e $vf/address ]] && [[ -L $vf/device ]] || continue
```
Comment (adrianchiris) [inline ubuntu/entrypoint.sh:94]: is this check necessary ? you can rely on the layout of sysfs.

File: ubuntu/entrypoint.sh (hunk 4 above)
Comment (ykulazhenkov) [reply ubuntu/entrypoint.sh]: Thanks. I removed this logic from PR. I will debug this issue and submit separate PR if required.

File: ubuntu/entrypoint.sh (hunk 1 above)
Comment (ykulazhenkov) [reply ubuntu/entrypoint.sh:130]: done

File: ubuntu/entrypoint.sh (hunk 2 above)
Comment (ykulazhenkov) [reply ubuntu/entrypoint.sh]: done

File: ubuntu/entrypoint.sh (hunk 3 above)
Comment (ykulazhenkov) [reply ubuntu/entrypoint.sh]: fixed

File: ubuntu/entrypoint.sh (hunk 5 above)
Comment (ykulazhenkov) [reply ubuntu/entrypoint.sh:173]: done

File: ubuntu/entrypoint.sh (hunk 6 above)
Comment (ykulazhenkov) [reply ubuntu/entrypoint.sh]: Yes, I think we can safely drop `[[ -e $vf/address ]] && [[ -L $vf/device ]]` check. Done

**Pull request merged**